dev-tools/omdb/src/bin/omdb/db.rs

```diff
@@ -5614,6 +5685,400 @@ impl From<&'_ Migration> for MigrationVmms {
+            let &Vmm {
+                id,
+                time_created: _,
+                time_deleted: _,
+                instance_id,
+                sled_id,
+                propolis_ip: _,
+                propolis_port: _,
+                runtime:
+                    db::model::VmmRuntimeState {
+                        state,
+                        r#gen,
+                        time_state_updated: _,
+                    },
+            } = vmm;
+            let sled = match sled {
+                Some(sled) => sled.serial_number(),
+                None => {
+                    eprintln!("WARN: no sled found with ID {sled_id}");
+                    "<unknown>"
+                }
+            };
+            VmmRow {
+                instance_id,
+                state: VmmStateRow { id, state, generation: r#gen.0.into() },
+                sled,
+            }
+        }
+    }
+
+    #[derive(Tabled)]
+    #[tabled(rename_all = "SCREAMING_SNAKE_CASE")]
+    struct VerboseVmmRow<'a> {
+        #[tabled(inline)]
+        inner: VmmRow<'a>,
+        sled_id: Uuid,
+        address: std::net::SocketAddr,
+        #[tabled(display_with = "datetime_rfc3339_concise")]
+        time_created: DateTime<Utc>,
+        #[tabled(display_with = "datetime_rfc3339_concise")]
+        time_updated: DateTime<Utc>,
+    }
+
+    impl<'a> From<&'a (Vmm, Option<Sled>)> for VerboseVmmRow<'a> {
+        fn from(it: &'a (Vmm, Option<Sled>)) -> Self {
+            let Vmm {
+                time_created,
+                time_deleted: _,
+                sled_id,
+                propolis_ip,
+                propolis_port,
+                ref runtime,
+                ..
+            } = it.0;
+            VerboseVmmRow {
+                sled_id,
+                inner: VmmRow::from(it),
+                address: std::net::SocketAddr::new(
+                    propolis_ip.ip(),
+                    propolis_port.into(),
+                ),
+                time_created,
+                time_updated: runtime.time_state_updated,
+            }
+        }
+    }
+
+    #[derive(Tabled)]
+    #[tabled(rename_all = "SCREAMING_SNAKE_CASE")]
+    struct WithDeleted<T: Tabled> {
+        #[tabled(inline)]
+        inner: T,
+        #[tabled(display_with = "datetime_opt_rfc3339_concise")]
+        time_deleted: Option<DateTime<Utc>>,
+    }
+
+    impl<'a, T> From<&'a (Vmm, Option<Sled>)> for WithDeleted<T>
+    where
+        T: Tabled + From<&'a (Vmm, Option<Sled>)>,
+    {
+        fn from(it: &'a (Vmm, Option<Sled>)) -> Self {
+            Self { inner: T::from(it), time_deleted: it.0.time_deleted }
+        }
+    }
+
+    let table = match (verbose, fetch_opts.include_deleted) {
+        (true, true) => tabled::Table::new(
+            vmms.iter().map(WithDeleted::<VerboseVmmRow>::from),
+        )
+        .with(tabled::settings::Style::empty())
+        .with(tabled::settings::Padding::new(0, 1, 0, 0))
+        .to_string(),
```

Comment:
I *think* that the `.with` + `.to_string()` type don't change the `tabled::Table` type, right?

If I'm reading this correctly, could we de-duplicate these lines across all cases of the match?

Reply:
...oh, I had stupidly assumed that `Table` was generic over the item type, so I thought the match couldn't return a `Table` because it would have a different type depending on the entry...but that's not actually the case, so we can make this a lot nicer! Thanks!